applepay/utils.py

```diff
@@ -66,3 +106,327 @@ def signing_time_is_valid(signature, current_time, threshold):
+    Returns:
+        object: the first filtered item from iterable or None
+            if no items matching the filter are found
+    """
+    filtered = ifilter(filter_func, iterable)
+    return next(filtered, None)
+
+
+def get_hashfunc_by_name(name, data):
+    """
+    Get a callable hashfunc by name.
+
+    This function can be used directly or with functools.partial, for example:
+
+    >>> hashfunc = functools.partial(get_hashfunc_by_name, 'sha256')
+    >>> hashfunc('sir robin').digest()
+
+    Args:
+        name (str): The string name of the desired algorithm
+        data (buffer): The buffer to hash
+    Returns:
+        callable: The callable hash function of the provided
+            algorithm updated with the data to be hashed
+    """
+    hashfunc = hashlib.new(name)
+    hashfunc.update(data)
+    return hashfunc
+
+
+def validate_message_digest(signed_attrs, hashed_payment_data):
+    """Validate the message_digest matches the provided payment.
+
+    Args:
+        signed_attrs (asn1crypto.cms.CMSAttributes): The signed
+            attributes from the signature
+        payment_data (str): the hashed payment data to validate
+            against
+    Returns:
+        boolean: True if the message digest matches the hashed
+            payment data, otherwise False.
+
+    """
+    message_digest_attr = get_first_from_iterable(
+        filter_func=lambda signed_attr: signed_attr['type'].dotted == payment.OID_MESSAGE_DIGEST,
+        iterable=signed_attrs
+    )
+
+    if not message_digest_attr:
+        logger.warning("No message digest found for the leaf cert.")
+        return False
+
+    message_digest = message_digest_attr['values'][0].native
+
+    if hashed_payment_data != message_digest:
+        logger.warning("Message digest does not match provided data.")
+        return False
+
+    return True
+
+
+def verify_signature(token, threshold=None):
+    """Verify the signature within an apple pay token according to
+    the #ApplePaySpec documentation.
+
+    Args:
+        token (dict): the PKPaymentToken object defined within the #ApplePaySpec
+    Kwargs:
+        threshold (timedelta, None): Amount of time to consider the token valid.
+            No validation will be performed if a threshold is not provided
+    Returns:
+        boolean: Indicates if the signature is valid or not
+    """
+
+    # We only bother to validate EC_v1.  RSA is only used
+    # for transactions from China and is not supported at this time.
+    if token['version'] != 'EC_v1':
+        logger.warning("Unsupported version {}".format(token['version']))
+        return False
+
+    # Extract and decode the signature object into a CMS object.
+    signature = token['signature']
+    signature_data = base64.b64decode(signature)
+    content_info = cms.ContentInfo.load(signature_data)
+    signed_data = content_info['content']
+    certificates = signed_data['certificates']
+
+    # There should be exactly 2 certificates present.
```

Comment:
Is this true? (are there always going to be exactly 2 certificates present for Applepay?)